crypto: Add new module openssl_pkcs12

gdelpierre opened the pull request:
##### SUMMARY
This module aims to allow a user to manage PKCS#12.
Internally it relies on the pyOpenSSL python library to interact with openssl.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
 - openssl_pkcs12

##### ANSIBLE VERSION
```
ansible 2.7
  config file = None
  configured module search path = [u'/Users/gdelpierre/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/gdelpierre/.local/lib/python2.7/site-packages/ansible
  executable location = /Users/gdelpierre/.local/bin/ansible
  python version = 2.7.15 (default, Jun 23 2018, 22:33:25) [GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)]
```

gdelpierre committed:
This commit changes:

  * [module] Create new module to handle PKCS#12 file.

gdelpierre commented:
cc @Spredzy @pilou-

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+        default: null
+        description:
+            - The PKCS#12 password.
+    privatekey:
```
`privatekey_path` is used in examples and code.

Spredzy replied:
s/privatekey/privatekey_path/ to match all the other module naming convention.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+        maciter_size=dict(default=1, type='int'),
+        mode=dict(default=0400, type='int'),
+        passphrase=dict(type='str', no_log=True),
+        path=dict(required=True, type='path'),
```
According to the documentation `aliases` keyword is missing for `path`.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+    force:
+        required: False
+        default: False
+        choices: [ True, False ]
```
There is no need to use `choices` for boolean value.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+        ca_certificates=dict(type='list'),
+        cert_path=dict(type='path'),
+        force=dict(default=False, type='bool'),
+        friendly_name=dict(required=True, type='str'),
```
According to the documentation `aliases` keyword is missing for `friendly_name`.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.crypto import OpenSSLModule, OpenSSLModuleError, load_certificate, load_privatekey
+from ansible.module_utils.pycompat24 import get_exception
```
There is no need to support Python 2.4, [see](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_4.rst#python-2-4-and-2-5-support-discontinuation).

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+        self.pkcs12 = crypto.PKCS12()
+
+        if self.ca_certificates:
+            ca_certs = tuple([load_certificate(ca_cert) for ca_cert in self.ca_certificates])
```
It seems `tuple` is superfluous and could be deleted.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+        except (IOError, OSError) as exc:
+            raise PkcsError(exc)
+
+    def check(self):
```
Is a `TODO` missing here ?

pilou- reviewed (commented):
Please, consider adding new tests within this pull request.

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+        choices: [ 'parse', 'export' ]
+        description:
+            - Create (export) or parse a PKCS #12.
+    dest:
```
s/dest/path, to match all the other module naming convention. I see no need for `dest`, `destfile` as the other modules do not use this.

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+    pyopenssl_found = True
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.crypto import OpenSSLModule, OpenSSLModuleError, load_certificate, load_privatekey
```
OpenSSLModule has been renamed OpenSSLObject
OpenSSLModuleError has been renamed OpenSSLObjectError

Also the other modules uses:

`from ansible.module_utils import crypto as crypto_utils`

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+from ansible.module_utils._text import to_native
+
+
+class PkcsError(OpenSSLModuleError):
```
s/OpenSSLModuleError/crypto_utils.OpenSSLObjectError/

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+    pass
+
+
+class Pkcs(OpenSSLModule):
```
s/OpenSSLModule/crypto_utils.OpenSSLObject/

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+class Pkcs(OpenSSLModule):
+
+    def __init__(self, module):
+        super(Pkcs, self).__init__(module)
```
Use the parent object. Look at current `openssl_privatekey`, `openssl_publickey` and `openssl_csr` implementation.

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+    def export(self):
+        ''' Generate pkcs#12 file archive. '''
+
+        if os.path.exists(self.path) and not self.force:
```
Use the check function as the criteria. see `openssl_privatekey`

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+                        self.iter_size,
+                        self.maciter_size
+                    )
+                )
```
Please ensure the file_common_args are applied here. see `openssl_privatekey`

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+                                        )
+
+        try:
+            with open(self.path, 'w') as archive:
```
s/w/wb for python3 compatibility

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+    def dump(self):
+        result = {
+            'changed': self.changed,
+            'path': self.path,
```
s/path/filename as the key of the result `dict()`

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+        argument_spec = argument_spec,
+        required_together = required_together,
+        supports_check_mode = True,
+        add_file_common_args = True,
```
Also have a reguired_if ('state', 'present', ['privatekey_path'])

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+    else:
+        if module.check_mode:
+            result = pkcs12.dump()
+            result['changed'] = pkcs12.check()
```
s/pcks12.check()/os.path.exists(module.params['path'])/

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,261 @@
+
+    if not pyopenssl_found:
+        module.fail_json(msg='The python pyOpenSSL library is required')
+
```
Missing a check here if the path directory actually exists

```
    base_dir = os.path.dirname(module.params['path'])
    if not os.path.isdir(base_dir):
        module.fail_json(
            name=base_dir,
            msg='The directory %s does not exist or the file is not a directory' % base_dir
        )
```

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,333 @@
+        required: False
+        description:
+            - List of CA certificate to include.
+    certificate:
```
s/certificate/cert_path/

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,333 @@
+        description:
+            - Should the file be regenerated even it it already exists.
+    friendly_name:
+        required: True
```
s/True/False

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,333 @@
+        description:
+            - The PKCS#12 password.
+    privatekey_path:
+        required: True
```
s/True/False

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,333 @@
+        ca_certificates=dict(type='list'),
+        cert_path=dict(type='path'),
+        force=dict(default=False, type='bool'),
+        friendly_name=dict(required=False, type='str', aliases=['name']),
```
No need to specify required if False

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,333 @@
+        maciter_size=dict(default=1, type='int'),
+        passphrase=dict(type='str', no_log=True),
+        path=dict(required=True, type='path'),
+        privatekey_path=dict(required=False, type='path'),
```
No need to specify required if False

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,333 @@
+
+        result = {
+            'changed': self.changed,
+            'filename': self.path,
```
Could you add one row here so we know which privatekey has been used for that (if we register the module output)

Like this: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/crypto/openssl_publickey.py#L235

Spredzy commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,394 @@
+
+        file_args = module.load_file_common_arguments(module.params)
+        if module.set_fs_attributes_if_different(file_args, False):
+            module.set_mode_if_different(self.path, this_mode, False)
```
No need to call explicitly this, it is already called in the `module.set_fs_attributes_if_different()` call [1]

[1] https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L1454-L1456

gdelpierre commented:
Sorry for the noise, my rebase goes wrong.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,365 @@
+        default: 'present'
+        choices: ['present', 'absent']
+        description:
+            - Whether the file should exist or not.
```
All parameters exept `path` are ignored when `state` is `absent`, this could be mentioned here.

pilou- replied:
Done

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,365 @@
+        """Serialize the object into a dictionary."""
+
+        result = {
+            'changed': self.changed,
```
Not sure why changed is added here, it could make sense to set value of `changed` key [here](https://github.com/ansible/ansible/pull/27320/files#diff-15692372a20724add1fc115ef36deff9R360).

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,365 @@
+
+    def __init__(self, module):
+        super(Pkcs, self).__init__(
+            module.params['path'],
```
`state` and `force` must be swapped.

pilou- replied:
Please, update the code (DONE)

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,365 @@
+
+        state_and_perms = super(Pkcs, self).check(module, perms_required)
+
+        def _check_pkey_passphrase():
```
Not sure if the nested method make it more readable.

Edited: nothing to change there.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,365 @@
+                                      self.mode)
+                os.write(pkcs12_file, self.pkcs12.export(self.passphrase, self.iter_size, self.maciter_size))
+                os.close(pkcs12_file)
+                self.changed = True
```
`self.changed` is already set to `True` by previous call to `self.remove()`.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,365 @@
+                raise PkcsError(exc)
+
+        file_args = module.load_file_common_arguments(module.params)
+        if module.set_fs_attributes_if_different(file_args, False):
```
`self.changed` should be used instead of `False`

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,365 @@
+                                      self.mode)
+                os.write(pkcs12_file, '%s%s' % (pkey, crt))
+                os.close(pkcs12_file)
+                self.changed = True
```
`self.changed` is already set to `True` by previous call to `self.remove()`.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,365 @@
+            try:
+                self.remove()
+
+                p12 = crypto.load_pkcs12(open(self.src, 'rb').read(),
```
file is not closed

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,365 @@
+                raise PkcsError(exc)
+
+        file_args = module.load_file_common_arguments(module.params)
+        if module.set_fs_attributes_if_different(file_args, False):
```
`self.changed` should be used instead of `False`

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,365 @@
+            if module.params['action'] == 'export':
+                pkcs12.generate(module)
+            else:
+                pkcs12.parse(module)
```
the last three common lines of `generate` and `parse` could be moved here.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,365 @@
+    def generate(self, module):
+        """Generate PKCS#12 file archive."""
+
+        if not self.check(module, perms_required=False) or module.params['force']:
```
use `self.force` instead of `module.params['force']`

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,365 @@
+    def parse(self, module):
+        """Read PKCS#12 file."""
+
+        if not self.check(module, perms_required=False) or module.params['force']:
```
use `self.force` instead of `module.params['force']`

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,365 @@
+            module.exit_json(**result)
+
+        try:
+            if module.params['action'] == 'export':
```
the first line of `generate` and `parse` could be moved here.

MarkusTeufelberger commented:
Since the bot mentioned me: I'll let @Spredzy and @pilou- do the reviews. If you need another shipit, feel free to comment.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,363 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
```
Add a copyright line, something like:
```
# Copyright (c) 2017 Guillaume Delpierre <gde@llew.me>
```

dagwieers replied:
Don't ask me why, but the used syntax includes a colon:
```bash
# Copyright: (c) 2017 Guillaume Delpierre <gde@llew.me>
```

pilou- replied:
This is no colon in the format described in [documentation](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/dev_guide/developing_modules_documenting.rst#copyright).

dagwieers replied:
Ah ? Most of the modules were modified with something like this:
https://github.com/ansible/ansible/commit/bf54a0c3e54ab2c2ed4e50c7b2c0ef5837fd69fc 

Which *does* include the colon...
cc: @Akasurde @abadger

abadger replied:
As long as the important pieces are all there, the exact format doesn't matter.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,363 @@
+---
+module: openssl_pkcs12
+author: "Guillaume Delpierre (@gdelpierre)"
+version_added: "2.4"
```
Due to the the [freeze](https://github.com/ansible/ansible/pull/27320#pullrequestreview-52365848), it will be merged in `2.5`.

mscherer commented:
-label c:cli/adhoc
-label c:cli/console
-label c:cli/doc
-label c:cli/galaxy
-label c:cli/pull
-label c:cli/vault
-label c:compat/tests
-label c:constants
-label c:errors/
-label c:executor/module_common
-label c:executor/play_iterator
-label c:executor/playbook_executor
-label c:executor/process/worker
-label c:executor/stats
-label c:executor/task_executor
-label c:executor/task_queue_manager
-label c:galaxy/
-label c:module_utils/basic
-label c:module_utils/facts
-label c:module_utils/urls
-label cloud
-label inventory
-label openstack

pilou- commented on test/integration/targets/openssl_pkcs12/aliases:
```diff
@@ -0,0 +1 @@
+posix/ci/group1
```
`destructive` and `needs/root` must be added because `setup_openssl` use `package` module.

* [X] Done

pilou- commented on test/integration/targets/openssl_pkcs12/tasks/main.yml:
```diff
@@ -0,0 +1,34 @@
+      state: absent
+      path: '{{ output_dir }}/ansible.p12'
+
+  when: pyopenssl_version.stdout|version_compare('16.0.0', '>=')
```
Add a comment explaining this requirement.

pilou- commented on test/integration/targets/openssl_pkcs12/tasks/main.yml:
```diff
@@ -0,0 +1,34 @@
+
+  - import_tasks: ../tests/validate.yml
+
+  - name: 'Delete PKCS#12 file'
```
This cleaning task must be moved in a `always` block.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,364 @@
+        module.fail_json(
+            name=base_dir,
+            msg='The directory %s does not exist or '
+                'the file is not a directory' % base_dir
```
Maybe `the path is not a directory` ?

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/35973/6/tests

pilou- commented:
<pre>
2017-09-01 15:10:57 TASK [openssl_pkcs12 : Validate PKCS#12] ***************************************
2017-09-01 15:10:58 fatal: [testhost]: FAILED! => {"changed": false, "failed": true, "msg": "The pexpect python module is required"}
2017-09-01 15:10:58 
2017-09-01 15:10:58 TASK [openssl_pkcs12 : Delete PKCS#12 file] ************************************
2017-09-01 15:10:59 fatal: [testhost]: FAILED! => {"changed": false, "failed": true, "msg": "action is export but the following are missing: friendly_name"}
</pre>
Two problems:
- pexcept not available on osx/10.11
- when `state` is `absent`, then `action` has the default value `export`, then `friendly_name` is required

mattclay commented:
`pexpect` should be installed using the `pip` module with `state: present` at the start of the test.

Spredzy commented:
shipit

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,371 @@
+        default: False
+        type: bool
+        description:
+            - Should the file be regenerated even it it already exists.
```
s/it it/if it/

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,371 @@
+        choices: ['present', 'absent']
+        description:
+            - Whether the file should exist or not.
+              All parameters except C(path) are ignored when state is absent.
```
Formatting function could be used here: `C(absent)`.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,371 @@
+    src:
+        description:
+            - PKCS#12 file path to parse.
+'''
```
Should not the `files` documentation fragment be used here ?

Edited: it seems only `mode` is used here, you could import `FILE_COMMON_ARGUMENTS` defined in basic.py, then reuse only `FILE_COMMON_ARGUMENTS['mode']` instead of all files arguments.

* [X] Done (all file common arguments used)

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,371 @@
+        self.src = module.params['src']
+        self.mode = module.params['mode']
+        if not self.mode:
+            self.mode = int('0400', 8)
```
`0o400` could be used here (Python 2.6 [handles](https://docs.python.org/3/whatsnew/2.6.html#pep-3127-integer-literal-support-and-syntax) this syntax).

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,371 @@
+                try:
+                    crypto_utils.load_privatekey(self.path,
+                                                 self.privatekey_passphrase)
+                    return True
```
This statement (`return True`) could be removed.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,371 @@
+                                          p12.get_privatekey())
+            crt = crypto.dump_certificate(crypto.FILETYPE_PEM,
+                                          p12.get_certificate())
+            os.close(p12)
```
I don't understand how `os.close(p12)` could work ? I mean: `p12` isn't a file descriptor, i tested locally:
```
In [1]: os.close(crypto.load_pkcs12(open('/tmp/test.p12', 'rb').read(), ''))
TypeError: an integer is required
```
If something need to be modified, could you add a test for that ?

pilou- reviewed (commented):
Maybe one problem with os.close call, FILE_COMMON_ARGUMENTS['mode'] could be used instead of adding all files arguments.

Commit:
fix tests

fix(lib/ansible/modules/crypto/openssl_pkcs12.py): pep8 + ansible-test
fix(test/integration/targets/openssl_pkcs12/aliases): missing alias
fix(test/integration/targets/openssl_pkcs12/tasks/main.yml): filters

Commit:
fix(openssl_pkcs12.py): pilou- review

Commit:
fix(openssl_pkcs12.py): pilou- review

pilou- commented on test/integration/targets/openssl_pkcs12/aliases:
```diff
@@ -0,0 +1,4 @@
+destructive
+needs/root
+posix/ci/group1
+shippable/posix/group1
```
Meanwhile, `posix/ci/group1` becomes `shippable/posix/group1`

* [X] Done

pilou- commented on test/integration/targets/openssl_pkcs12/tests/validate.yml:
```diff
@@ -0,0 +1,11 @@
+- name: 'Validate PKCS#12'
+  shell: "openssl pkcs12 -info -in {{ output_dir }}/ansible.p12 -nodes -passin pass:''"
```
`command` can be used here.

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,371 @@
+              All parameters except C(path) are ignored when state is C(absent).
+    src:
+        description:
+            - PKCS#12 file path to parse.
```
add `extends_documentation_fragment: [files]` (because `add_file_common_args` is enabled) ([example](https://github.com/ansible/ansible/blob/d63ba5999b7936a96ee000244b45151e75a116ec/lib/ansible/modules/net_tools/basics/get_url.py#L156)). Then doc related to `mode` parameter could be removed.

* [X] Done

pilou- commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -0,0 +1,371 @@
+    result = pkcs12.dump()
+    result['changed'] = changed
+    if os.path.exists(module.params['path']):
+        file_mode = oct(os.stat(module.params['path']).st_mode)[-4:]
```
Another way `"%04o" % stat.S_IMODE(os.stat(module.params['path']).st_mode)` (no change required here)

Commit:
Pilou review

Commit:
tests: Install pexpect

pilou- reviewed (commented):
Could you add a test with ~`action=export`~ `action=parse` ? It will allow to test change about `os.close()` ([ref](https://github.com/ansible/ansible/pull/27320#discussion_r143666205))

* [X] Done

Commit:
add test with action: parse

pilou- reviewed (approved):
LGTM

abadger commented:
Merged for inclusion in 2.7.0.  Thanks @gdelpierre  for writing the module and @Spredzy and @pilou- for the review!
